MatrixKit/Categories/NSString+MatrixKit.swift

```diff
@@ -0,0 +1,58 @@
+// 
+// Copyright 2020 The Matrix.org Foundation C.I.C
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+import MatrixSDK.MXLog
+
+public extension NSString {
+    /// Gets the first URL contained in the string.
+    /// - Returns: A URL if detected, otherwise nil.
+    @objc func mxk_firstURLDetected() -> NSURL? {
+        guard let linkDetector = try? NSDataDetector(types: NSTextCheckingResult.CheckingType.link.rawValue) else {
+            MXLog.debug("[NSString+URLDetector]: Unable to create link detector.")
+            return nil
+        }
+        
+        var detectedURL: NSURL?
+        
+        // enumerate all urls that were found in the string to ensure
+        // detection of a valid link if there are invalid links preceding it
+        linkDetector.enumerateMatches(in: self as String,
+                                      options: [],
+                                      range: NSRange(location: 0, length: self.length)) { match, flags, stop in
+            guard let match = match else { return }
+            
+            // check if the match is a valid url
+            let urlString = self.substring(with: match.range)
+            guard let url = NSURL(string: urlString) else { return }
+            
+            // ensure the match is a web link
+            guard let scheme = url.scheme?.lowercased(),
+                  scheme == "https" || scheme == "http"
+            else { return }
+            
+            // discard any matrix.to links
+            guard let host = url.host?.lowercased(),
+                  host != "matrix.to"
```

Comment:
We should had or use constants like `kMXMatrixDotToUrl`. But probably better to add a parameters like `hostsToIgnore: [String]`. From the app you can decide which URLs to ignore. A default implementation can use something like: `hostsToIgnore: [URL(string: kMXMatrixDotToUrl.host)]` (make null safety checks before).

Reply:
Updated, which makes the function more generally usable, although I just realised doesn't allow customisation in `MXKRoomBubbleComponent`'s [detection](https://github.com/matrix-org/matrix-ios-kit/pull/893/files#diff-b320829d3dedae327a784fc36287f2698f84ed60639b987ce124aea30f26c7e7R128).

Wondering if it would make more sense to add a `firstURLDetectionIgnoredHosts` setting to `MXKAppSetting` and use that when no arguments are passed to `mxk_firstURLDetected()`.

Reply:
It worth to keep the method with open parameters for test and reuse. We can keep `firstURLDetectionIgnoredHosts` as is and `mxk_firstURLDetected()` can use `MXKAppSetting.urlPreviewIgnoredHosts` for example

Reply:
Updated, I think it should make sense.